component/pyroscope/ebpf/ebpfspy/session.go

```diff
@@ -0,0 +1,325 @@
+	}
+	var sfs []sf
+	knownStacks := map[uint32]bool{}
+	for i := range keys {
+		ck := &keys[i]
+		value := values[i]
+
+		if ck.UserStack >= 0 {
+			knownStacks[uint32(ck.UserStack)] = true
+		}
+		if ck.KernStack >= 0 {
+			knownStacks[uint32(ck.KernStack)] = true
+		}
+		labels := s.targetFinder.FindTarget(ck.Pid)
+		if labels == nil {
+			continue
+		}
+		uStack := s.getStack(ck.UserStack)
+		kStack := s.getStack(ck.KernStack)
+		sfs = append(sfs, sf{
+			pid:    ck.Pid,
+			uStack: uStack,
+			kStack: kStack,
+			count:  value,
+			comm:   getComm(ck),
+			labels: labels,
+		})
+	}
+
+	sb := stackBuilder{}
+	for _, it := range sfs {
+		sb.rest()
+		sb.append(it.comm)
+		if s.options.CollectUser {
+			s.walkStack(&sb, it.uStack, it.pid)
+		}
+		if s.options.CollectKernel {
+			s.walkStack(&sb, it.kStack, 0)
+		}
+		lo.Reverse(sb.stack)
+		cb(it.labels, sb.stack, uint64(it.count), it.pid)
+	}
+	if err = s.clearCountsMap(keys, batch); err != nil {
+		return fmt.Errorf("clear counts map %w", err)
+	}
+	if err = s.clearStacksMap(knownStacks); err != nil {
+		return fmt.Errorf("clear stacks map %w", err)
+	}
+	return nil
+}
+
+func (s *session) Stop() {
+	for _, pe := range s.perfEvents {
+		_ = pe.Close()
+	}
+	s.perfEvents = nil
+	s.bpf.Close()
+}
+
+func (s *session) Update(options SessionOptions) error {
+	s.symCache.UpdateOptions(options.CacheOptions)
+	err := s.updateSampleRate(options.SampleRate)
+	if err != nil {
+		return err
+	}
+	s.options = options
+	return nil
+}
+
+func (s *session) DebugInfo() interface{} {
+	return SessionDebugInfo{
+		ElfCache: s.symCache.ElfCacheDebugInfo(),
+		PidCache: s.symCache.PidCacheDebugInfo(),
+	}
+}
+
+func (s *session) initArgs() error {
+	var zero uint32
+	var tgidFilter uint32
+	if s.pid <= 0 {
+		tgidFilter = 0
+	} else {
+		tgidFilter = uint32(s.pid)
+	}
+	arg := &profileBssArg{
+		TgidFilter: tgidFilter,
+	}
+	if err := s.bpf.Args.Update(&zero, arg, 0); err != nil {
+		return fmt.Errorf("init args fail: %w", err)
+	}
+	return nil
+}
+
+func (s *session) attachPerfEvents() error {
+	var cpus []uint
+	var err error
+	if cpus, err = cpuonline.Get(); err != nil {
+		return fmt.Errorf("get cpuonline: %w", err)
+	}
+	for _, cpu := range cpus {
+		pe, err := newPerfEvent(int(cpu), s.options.SampleRate)
+		if err != nil {
+			return fmt.Errorf("new perf event: %w", err)
+		}
+		s.perfEvents = append(s.perfEvents, pe)
+
+		err = pe.attachPerfEvent(s.bpf.profilePrograms.DoPerfEvent)
+		if err != nil {
+			return fmt.Errorf("attach perf event: %w", err)
+		}
+	}
+	return nil
+}
+
+func (s *session) getStack(stackId int64) []byte {
+	if stackId < 0 {
+		return nil
+	}
+	stackIdU32 := uint32(stackId)
+	res, err := s.bpf.profileMaps.Stacks.LookupBytes(stackIdU32)
+	if err != nil {
+		return nil
+	}
+	return res
+}
+
+func (s *session) walkStack(sb *stackBuilder, stack []byte, pid uint32) {
+	if len(stack) == 0 {
+		return
+	}
+	var stackFrames []string
+	for i := 0; i < 127; i++ {
+		it := stack[i*8 : i*8+8]
```

Comment:
Would be great to explain/comment the layout of `stack []byte` it and ip doesn't ring a bell to me.